File: Application/Module/CRM/Objects/Location.php
<?php

/**
 * Location
 *
 * Data access layer for the locations table.
 * Locations are always scoped to an account_id.
 */
class Location
{
    private DB $db;

    const LOCATION_TYPES = ['Bill To', 'Ship To'];

    const STATUSES = ['Active', 'Inactive', 'Closed', 'Temporary'];

    const VALIDATION_STATUSES = ['Verified', 'Pending', 'Invalid'];

    public function __construct(DB $db)
    {
        $this->db = $db;
    }

    // =========================================================================
    // Queries
    // =========================================================================

    public function findByAccount(int $accountId): array
    {
        return $this->db->query(
            'SELECT * FROM locations
              WHERE account_id = ?
              ORDER BY is_primary DESC, location_name ASC, id ASC',
            [$accountId]
        );
    }

    public function findById(int $locationId, int $accountId): ?array
    {
        return $this->db->queryOne(
            'SELECT * FROM locations WHERE id = ? AND account_id = ?',
            [$locationId, $accountId]
        );
    }

    public function countByAccount(int $accountId): int
    {
        $row = $this->db->queryOne(
            'SELECT COUNT(*) AS n FROM locations WHERE account_id = ?',
            [$accountId]
        );
        return (int) ($row['n'] ?? 0);
    }

    // =========================================================================
    // Mutations
    // =========================================================================

    private function sanitize(array $data): array
    {
        return [
            'location_name'             => trim($data['location_name']             ?? ''),
            'location_type'             => in_array($data['location_type'] ?? '', self::LOCATION_TYPES, true)
                                            ? $data['location_type'] : '',
            'location_status'           => in_array($data['location_status'] ?? '', self::STATUSES, true)
                                            ? $data['location_status'] : 'Active',
            'is_primary'                => !empty($data['is_primary']) ? 1 : 0,
            'validation_status'         => in_array($data['validation_status'] ?? '', self::VALIDATION_STATUSES, true)
                                            ? $data['validation_status'] : 'Pending',
            'street_address_1'          => trim($data['street_address_1']          ?? ''),
            'street_address_2'          => trim($data['street_address_2']          ?? ''),
            'street_address_3'          => trim($data['street_address_3']          ?? ''),
            'city'                      => trim($data['city']                      ?? ''),
            'state_province'            => trim($data['state_province']            ?? ''),
            'zip_postal_code'           => trim($data['zip_postal_code']           ?? ''),
            'country_region'            => trim($data['country_region']            ?? ''),
            'county'                    => trim($data['county']                    ?? ''),
            'district_neighborhood'     => trim($data['district_neighborhood']     ?? ''),
            'building_name_number'      => trim($data['building_name_number']      ?? ''),
            'floor_suite_apartment'     => trim($data['floor_suite_apartment']     ?? ''),
            'intersection_cross_street' => trim($data['intersection_cross_street'] ?? ''),
            'po_box'                    => trim($data['po_box']                    ?? ''),
            'latitude'                  => ($data['latitude']       ?? '') !== '' ? (float) $data['latitude']       : null,
            'longitude'                 => ($data['longitude']      ?? '') !== '' ? (float) $data['longitude']      : null,
            'timezone_utc_offset'       => trim($data['timezone_utc_offset']       ?? ''),
            'geofence_radius'           => ($data['geofence_radius'] ?? '') !== '' ? (int) $data['geofence_radius'] : null,
            'dock_instructions'         => trim($data['dock_instructions']         ?? ''),
            'receiving_hours'           => trim($data['receiving_hours']           ?? ''),
            'liftgate_required'         => !empty($data['liftgate_required'])  ? 1 : 0,
            'vehicle_clearance'         => trim($data['vehicle_clearance']         ?? ''),
            'forklift_available'        => !empty($data['forklift_available']) ? 1 : 0,
            'gate_entry_code'           => trim($data['gate_entry_code']           ?? ''),
            'preferred_carrier'         => trim($data['preferred_carrier']         ?? ''),
        ];
    }

    public function add(int $accountId, array $data): array
    {
        $d = $this->sanitize($data);

        $id = $this->db->insert(
            "INSERT INTO locations (
                account_id,
                location_name, location_type, location_status, is_primary, validation_status,
                street_address_1, street_address_2, street_address_3,
                city, state_province, zip_postal_code, country_region,
                county, district_neighborhood, building_name_number,
                floor_suite_apartment, intersection_cross_street, po_box,
                latitude, longitude, timezone_utc_offset, geofence_radius,
                dock_instructions, receiving_hours, liftgate_required,
                vehicle_clearance, forklift_available, gate_entry_code, preferred_carrier
             ) VALUES (
                ?,
                ?, ?, ?, ?, ?,
                ?, ?, ?,
                ?, ?, ?, ?,
                ?, ?, ?,
                ?, ?, ?,
                ?, ?, ?, ?,
                ?, ?, ?,
                ?, ?, ?, ?
             )",
            [
                $accountId,
                $d['location_name'], $d['location_type'], $d['location_status'],
                $d['is_primary'], $d['validation_status'],
                $d['street_address_1'], $d['street_address_2'], $d['street_address_3'],
                $d['city'], $d['state_province'], $d['zip_postal_code'], $d['country_region'],
                $d['county'], $d['district_neighborhood'], $d['building_name_number'],
                $d['floor_suite_apartment'], $d['intersection_cross_street'], $d['po_box'],
                $d['latitude'], $d['longitude'], $d['timezone_utc_offset'], $d['geofence_radius'],
                $d['dock_instructions'], $d['receiving_hours'], $d['liftgate_required'],
                $d['vehicle_clearance'], $d['forklift_available'], $d['gate_entry_code'],
                $d['preferred_carrier'],
            ]
        );

        return ['ok' => true, 'id' => (int) $id, 'error' => null];
    }

    public function update(int $locationId, int $accountId, array $data): array
    {
        $d = $this->sanitize($data);

        $affected = $this->db->execute(
            "UPDATE locations SET
                location_name = ?, location_type = ?, location_status = ?,
                is_primary = ?, validation_status = ?,
                street_address_1 = ?, street_address_2 = ?, street_address_3 = ?,
                city = ?, state_province = ?, zip_postal_code = ?, country_region = ?,
                county = ?, district_neighborhood = ?, building_name_number = ?,
                floor_suite_apartment = ?, intersection_cross_street = ?, po_box = ?,
                latitude = ?, longitude = ?, timezone_utc_offset = ?, geofence_radius = ?,
                dock_instructions = ?, receiving_hours = ?, liftgate_required = ?,
                vehicle_clearance = ?, forklift_available = ?, gate_entry_code = ?,
                preferred_carrier = ?
             WHERE id = ? AND account_id = ?",
            [
                $d['location_name'], $d['location_type'], $d['location_status'],
                $d['is_primary'], $d['validation_status'],
                $d['street_address_1'], $d['street_address_2'], $d['street_address_3'],
                $d['city'], $d['state_province'], $d['zip_postal_code'], $d['country_region'],
                $d['county'], $d['district_neighborhood'], $d['building_name_number'],
                $d['floor_suite_apartment'], $d['intersection_cross_street'], $d['po_box'],
                $d['latitude'], $d['longitude'], $d['timezone_utc_offset'], $d['geofence_radius'],
                $d['dock_instructions'], $d['receiving_hours'], $d['liftgate_required'],
                $d['vehicle_clearance'], $d['forklift_available'], $d['gate_entry_code'],
                $d['preferred_carrier'],
                $locationId, $accountId,
            ]
        );

        return $affected > 0
            ? ['ok' => true,  'error' => null]
            : ['ok' => false, 'error' => 'Location not found.'];
    }

    public function remove(int $locationId, int $accountId): bool
    {
        $affected = $this->db->execute(
            'DELETE FROM locations WHERE id = ? AND account_id = ?',
            [$locationId, $accountId]
        );
        return $affected > 0;
    }
}

File: iPhone/AplosCRM/Sources/AplosCRM/Location.swift
import Foundation

enum LocationEnums {
    static let types = ["Bill To", "Ship To"]
    static let statuses = ["Active", "Inactive", "Closed", "Temporary"]
    static let validationStatuses = ["Verified", "Pending", "Invalid"]
}

struct Location: Identifiable, Decodable {
    let id: Int
    let accountID: Int
    let locationName: String?
    let locationType: String?
    let locationStatus: String?
    let isPrimary: Int?
    let validationStatus: String?
    let streetAddress1: String?
    let streetAddress2: String?
    let streetAddress3: String?
    let city: String?
    let stateProvince: String?
    let zipPostalCode: String?
    let countryRegion: String?
    let county: String?
    let districtNeighborhood: String?
    let buildingNameNumber: String?
    let floorSuiteApartment: String?
    let intersectionCrossStreet: String?
    let poBox: String?
    let latitude: String?
    let longitude: String?
    let timezoneUTCOffset: String?
    let geofenceRadius: Int?
    let dockInstructions: String?
    let receivingHours: String?
    let liftgateRequired: Int?
    let vehicleClearance: String?
    let forkliftAvailable: Int?
    let gateEntryCode: String?
    let preferredCarrier: String?

    enum CodingKeys: String, CodingKey {
        case id, city, county, latitude, longitude
        case accountID = "account_id"
        case locationName = "location_name"
        case locationType = "location_type"
        case locationStatus = "location_status"
        case isPrimary = "is_primary"
        case validationStatus = "validation_status"
        case streetAddress1 = "street_address_1"
        case streetAddress2 = "street_address_2"
        case streetAddress3 = "street_address_3"
        case stateProvince = "state_province"
        case zipPostalCode = "zip_postal_code"
        case countryRegion = "country_region"
        case districtNeighborhood = "district_neighborhood"
        case buildingNameNumber = "building_name_number"
        case floorSuiteApartment = "floor_suite_apartment"
        case intersectionCrossStreet = "intersection_cross_street"
        case poBox = "po_box"
        case timezoneUTCOffset = "timezone_utc_offset"
        case geofenceRadius = "geofence_radius"
        case dockInstructions = "dock_instructions"
        case receivingHours = "receiving_hours"
        case liftgateRequired = "liftgate_required"
        case vehicleClearance = "vehicle_clearance"
        case forkliftAvailable = "forklift_available"
        case gateEntryCode = "gate_entry_code"
        case preferredCarrier = "preferred_carrier"
    }
}

struct LocationsResponse: Decodable {
    let ok: Bool
    let data: [Location]
}

struct LocationDetailResponse: Decodable {
    let ok: Bool
    let data: Location
}

struct LocationRequest: Encodable {
    var locationName: String
    var locationType: String?
    var locationStatus: String?
    var isPrimary: Int?
    var validationStatus: String?
    var streetAddress1: String?
    var streetAddress2: String?
    var city: String?
    var stateProvince: String?
    var zipPostalCode: String?
    var countryRegion: String?

    enum CodingKeys: String, CodingKey {
        case city
        case locationName = "location_name"
        case locationType = "location_type"
        case locationStatus = "location_status"
        case isPrimary = "is_primary"
        case validationStatus = "validation_status"
        case streetAddress1 = "street_address_1"
        case streetAddress2 = "street_address_2"
        case stateProvince = "state_province"
        case zipPostalCode = "zip_postal_code"
        case countryRegion = "country_region"
    }

    init(from location: Location? = nil) {
        locationName = location?.locationName ?? ""
        locationType = location?.locationType
        locationStatus = location?.locationStatus ?? "Active"
        isPrimary = location?.isPrimary
        validationStatus = location?.validationStatus ?? "Pending"
        streetAddress1 = location?.streetAddress1
        streetAddress2 = location?.streetAddress2
        city = location?.city
        stateProvince = location?.stateProvince
        zipPostalCode = location?.zipPostalCode
        countryRegion = location?.countryRegion
    }
}
